server/src/main/java/com/broadcom/lsp/cobol/service/delegates/hover/VariableHover.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (c) 2020 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *    Broadcom, Inc. - initial API and implementation
+ *
+ */
+package com.broadcom.lsp.cobol.service.delegates.hover;
+
+import com.broadcom.lsp.cobol.core.model.Locality;
+import com.broadcom.lsp.cobol.core.model.variables.*;
+import com.broadcom.lsp.cobol.core.semantics.outline.RangeUtils;
+import com.broadcom.lsp.cobol.service.CobolDocumentModel;
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Singleton;
+import lombok.NonNull;
+import org.eclipse.lsp4j.*;
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * The class provides hover information for variables.
+ */
+@Singleton
+public class VariableHover implements HoverProvider {
+  @Nullable
+  @Override
+  public Hover getHover(@Nullable CobolDocumentModel document, @NonNull TextDocumentPositionParams position) {
+    if (document == null) return null;
+    Variable variable = findVariableByPosition(document.getAnalysisResult().getVariables(), position);
+    if (variable == null) return null;
+    String prefix = "";
+    List<String> lines = new ArrayList<>();
+    for (String parentLine: parentsDescriptions(variable.getParent())) {
+      lines.add(prefix + parentLine);
+      prefix += "  ";
+    }
+    lines.add(prefix + variable.getDisplayLine());
+    prefix += "  ";
+    for (String childLine: childrenDescriptions(variable))
+      lines.add(prefix + childLine);
+    return new Hover(ImmutableList.of(Either.forRight(
+        new MarkedString("cobol", String.join("\n", lines)))));
+  }
+
+  private static Variable findVariableByPosition(Collection<Variable> variables, TextDocumentPositionParams position) {
+    for (Variable variable: variables)
+      if (RangeUtils.isInside(position, variable.getDefinition().toLocation())
+          || variable.getUsages().stream()
+               .map(Locality::toLocation)
+               .anyMatch(location -> RangeUtils.isInside(position, location)))
+        return variable;
+    return null;
+  }
+
+  private static List<String> parentsDescriptions(Variable variable) {
+    if (variable == null) return new ArrayList<>();
```

Comment:
In below childrenDescriptions returned  Collections.emptyList(), I guess you will modify this later, right?

Reply:
And there are conflicts

Reply:
Yes, I call `parentsDescriptions` recursively and add parents into this array. And `childrenDescriptions` called only once.